oauthlib/oauth2/rfc6749/grant_types/openid_connect.py

```diff
@@ -76,6 +76,65 @@ def validate_authorization_request(self, request):
         return self._handler_for_request(request).validate_authorization_request(request)
 
 
+class ImplicitTokenGrantDispatcher(object):
+    """
+    This is an adapter class that will route simple Authorization Code requests, those that have response_type=code and a scope
+    including 'openid' to either the default_auth_grant or the oidc_auth_grant based on the scopes requested.
+    """
+    def __init__(self, default_implicit_grant=None, oidc_implicit_grant=None):
+        self.default_implicit_grant = default_implicit_grant
+        self.oidc_implicit_grant = oidc_implicit_grant
+
+    def _handler_for_request(self, request):
+        handler = self.default_implicit_grant
+
+        if request.scopes and "openid" in request.scopes and 'id_token' in request.response_type:
+            handler = self.oidc_implicit_grant
+
+        log.debug('Selecting handler for request %r.', handler)
+        return handler
+
+    def create_authorization_response(self, request, token_handler):
+        return self._handler_for_request(request).create_authorization_response(request, token_handler)
+
+    def validate_authorization_request(self, request):
+        return self._handler_for_request(request).validate_authorization_request(request)
+
+
+class AuthTokenGrantDispatcher(object):
+    """
+    This is an adapter class that will route simple Token requests, those that authorization_code have a scope
+    including 'openid' to either the default_token_grant or the oidc_token_grant based on the scopes requested.
+    """
+    def __init__(self, request_validator, default_token_grant=None, oidc_token_grant=None):
+        self.default_token_grant = default_token_grant
+        self.oidc_token_grant = oidc_token_grant
+        self.request_validator = request_validator
+
+    def _handler_for_request(self, request):
+        handler = self.default_token_grant
+        scopes = ()
+        parameters = dict(request.decoded_body)
+        client_id = parameters.get('client_id', None)
+        code = parameters.get('code', None)
+        redirect_uri = parameters.get('redirect_uri', None)
+
+        # If code is not pressent fallback to `default_token_grant` wich will
+        # raise an error for the missing `code` in `create_token_response` step.
+        if code:
+            scopes = self.request_validator.get_authorization_code_scopes(client_id, code, redirect_uri)
```

Comment:
What I don't particularly like, but also not really sure how to do differently, is that this forces to lookup the grant twice; once in `get_authorization_code_scopes()` and once in `validate_code()`. Not ideal for performance.

I wonder if we should pass in the `request` object here, as with many of the other validator methods. Then at least one can choose to attach the result of the first lookup to the request.

Reply:
This is a gap between stateless endpoints:

* Authorization Endpoint
* Token Endpoint

`request_validator.get_authorization_code_scopes()` is called before `request_validator.validate_code()` which is used inside `AuthorizationCodeGrant.validate_token_request()` we can call `request_validator.validate_code()` here and pass in `request.code_validate` a `True` value indicating the code was already validated and in `AuthorizationCodeGrant.validate_token_request()` line 410 add a check before call `request_validator.validate_code()` again.

Reply:
I guess what I meant was: Could we use the following signature instead?

    def get_authorization_code_scopes(self, client_id, code, redirect_uri, request):

This seems to be a common pattern already, and it would allow us to cache database lookups on the `request` object easily across validation calls.

Reply:
@skion done